src/sentry/reprocessing2.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import absolute_import
+
+import uuid
+import hashlib
+import logging
+import sentry_sdk
+
+from django.conf import settings
+
+from sentry import nodestore, features, eventstore
+from sentry.attachments import CachedAttachment, attachment_cache
+from sentry.models import EventAttachment
+from sentry.utils.cache import cache_key_for_event
+from sentry.eventstore.processing import event_processing_store
+
+logger = logging.getLogger("sentry.reprocessing")
+
+
+def _generate_unprocessed_event_node_id(project_id, event_id):
+    return hashlib.md5(
+        u"{}:{}:unprocessed".format(project_id, event_id).encode("utf-8")
+    ).hexdigest()
+
+
+def save_unprocessed_event(project, event_id):
+    """
+    Move event from event_processing_store into nodestore. Only call if event
+    has outcome=accepted.
+    """
+    if not features.has("projects:reprocessing-v2", project, actor=None):
+        return
+
+    with sentry_sdk.start_span(
+        op="sentry.reprocessing2.save_unprocessed_event.get_unprocessed_event"
+    ):
+        data = event_processing_store.get(
+            cache_key_for_event({"project": project.id, "event_id": event_id}), unprocessed=True
+        )
+        if data is None:
+            return
+
+    with sentry_sdk.start_span(op="sentry.reprocessing2.save_unprocessed_event.set_nodestore"):
+        node_id = _generate_unprocessed_event_node_id(project_id=project.id, event_id=event_id)
+        nodestore.set(node_id, data)
+
+
+def backup_unprocessed_event(project, data):
+    """
+    Backup unprocessed event payload into redis. Only call if event should be
+    able to be reprocessed.
+    """
+
+    if not features.has("projects:reprocessing-v2", project, actor=None):
+        return
+
+    event_processing_store.store(data, unprocessed=True)
+
+
+def delete_unprocessed_events(events):
+    node_ids = [
+        _generate_unprocessed_event_node_id(event.project_id, event.event_id) for event in events
+    ]
+    nodestore.delete_multi(node_ids)
+
+
+def reprocess_event(project_id, event_id, start_time):
+    node_id = _generate_unprocessed_event_node_id(project_id=project_id, event_id=event_id)
+
+    with sentry_sdk.start_span(op="reprocess_events.nodestore.get"):
+        data = nodestore.get(node_id)
+
+    from sentry.tasks.store import preprocess_event_from_reprocessing
+    from sentry.ingest.ingest_consumer import CACHE_TIMEOUT
+
+    # Take unprocessed data from old event and save it as unprocessed data
+    # under a new event ID. The second step happens in pre-process. We could
+    # save the "original event ID" instead and get away with writing less to
+    # nodestore, but doing it this way makes the logic slightly simpler.
+
+    # Step 1: Fix up the event payload for reprocessing and put it in event
+    # cache/event_processing_store
+    tags = dict(data.get("tags") or ())
+    orig_event_id = tags["original_event_id"] = data["event_id"]
```

Comment:
Putting this in tags is fine, as it gives us a simple way to search by it and jump to the original event. However, we might want to consider adding a dedicated attribute in the future similar to "release" so that we can apply better special handling.

Reply:
yeah this is an absolutely massive hack. this should really be its own snuba column but I was trying to avoid binding any time from S&S

Reply:
Agreed. As this remains feature flagged for a while, this gives us a chance to revise this. Let's follow up on this.